### DISCUSSION TITLE
To output the material properties on element-face for visualization

### DISCUSSION THREAD

**Discussion by schakrabortygithub** [idaholab/moose#26554] [Q&A General] [answered]
Hi,

I was wondering if MOOSE already has some option to output material properties on the element-faces for visualization. 

If material properties for some element faces are triggered to be computed, e.g, via boundary kernel or interface kernel, is it possible to output those values in exodus file for visualization? 
For example, when I output some material property(yy component of stress tensor for an elastic problem) in an auxvariable of type Constant-Monomial using auxkernel:

```
  [./Aux_pk2_yy]
   type = RankTwoAux
   variable = pk2_yy
   rank_two_tensor = stress
   index_j = 1
   index_i = 1
   #boundary = 'bottom '
  [../]
```
I get this:
<img width="550" alt="Screen Shot 2024-01-15 at 12 54 03 AM" src="https://github.com/idaholab/moose/assets/35871741/a2f306ca-06a2-40af-9caa-a52051495baa">>

And if I specify the `boundary = 'bottom '` optional parameter, I get this:
<img width="550" alt="Screen Shot 2024-01-15 at 12 54 11 AM" src="https://github.com/idaholab/moose/assets/35871741/1053ee25-b0b1-43f2-9b91-348357631b71">>

Specifying the 'boundary' optional parameter, it seems to still output the volumetric values for those elements that are adjacent to that boundary, though values are little off from the previous one.

And if I select a boundary-set(bottom) in paraview then those properties(pk2_yy) are not available to visualize at all.
Any suggestion?

Best
Subhendu

**Comment by GiudGiud** [comment, marked as answer]
Hello

The `boundary` parameter to an auxiliary kernel is to restrict output to dofs near the boundary (or on the boundary for a nodal variable).
It does not have a special behavior for those auxkernels meant to output material properties. 

It's not implemented. I think it can be implemented. I would create a new auxkernel `FaceRankTwoMatAux` that retrieves the face material property. I think I would use the same trick as for the `boundary` parameter, output the face value as a volumetric value in the element nearby, as we do not have much support for face variable output rn (though we have more and more face variable capability, so it s something that could be implemented too)

Guillaume

**Reply by schakrabortygithub** [reply]
Thanks @GiudGiud for the suggestion. I will try to implement `FaceRankTwoMatAux` to output face materials property.

**Reply by GiudGiud** [reply]
@lindsayad if you have some advice on face material property output

**Reply by lindsayad** [reply]
If you use the `boundary` parameter, you will get the face material property value. The issue with the visualization is that you need to color the internal values something, and since there is nothing setting the internal dof values, they are simply zero. Writing a new aux kernel would not do anything about the big red block for the dofs not on the boundary

**Reply by lindsayad** [reply]
Looking at your image again... the internal values in the bottom picture are not zero?! That is very troubling if you don't have some other aux kernel also outputting to the `pk2_yy` variable

**Reply by schakrabortygithub** [reply]
@lindsayad, the internal values in the second image is indeed zero. The confusion may be arising due to the range of the color bar. Since highest magnitude in the color bar is -1.8e+03, zero values are also getting colored with red.

When I specify the `boundary` parameter it does compute the surface-averaged magnitude of that material property but places those values as volumetric quantity corresponding to that 3D element. 
I was wondering if we can output those values as surface quantity corresponding to the face-elements of that 3D parent element. That would help to inspect those material properties without need to declare multiple auxvariables one for volumetric values and one for each boundaries(to avoid common element).

**Reply by lindsayad** [reply]
I'm still unclear on what you want. We could (potentially, it would be a fair amount of work) create a variable that lives on lower dimensional elements coincident with the higher-dimensional element boundary face, and then populate that variable with one aux kernel, and then populate another aux variable that has the volumetric evaluations. Alternatively, you could go down the custom aux kernel route like @GiudGiud suggested. You would want to set the `_dual_restrictable` parameter like is done in the `GapValueAux` and `NearestNodeValueAux` framework classes. And then for elements adjacent to your `boundary` you could do face-based evaluations, and then for elements not adjacent you could do volumetric evaluations. Note that this is what I expect for the outcome pursuing this path:

- for CONSTANT MONOMIAL variables, your visualization of boundary-adjacent elements will appear volumetric just by definition of the variable family type (it's a constant over the element)
- for higher order MONOMIAL, your evaluation of the boundary-adjacent element is actually rank deficient ... there is not enough information from the aux kernel to populate all of the element's dofs. We do a singular value decomposition in this case. All that you're ensured is that the evaluation of the variable on the face will correspond to your material property evaluation. I don't know what the variable will look like in other places of the element
- for nodal variables ... well these don't work with material properties so I don't need to say anything about them 😄

**Reply by schakrabortygithub** [reply]
@lindsayad,
Thanks for  the suggestions. Your first suggestion seems to be the ideal one, i.e, 'create a variable that lives on lower dimensional elements coincident with the higher-dimensional element boundary face'. But I do not need that one right now. I Just wanted to make sure that feature is not there in moose yet.

**Reply by lindsayad** [reply]
You should in theory be able to do this now from your end. You can use a `LowerDBlockFromSidesetGenerator` and then create an aux variable and aux kernel restricted to that block. I cannot guarantee that this will work with the face material property evaluation; if it doesn't it wouldn't be too difficult to add support for it. We do have support for it in other places

**Reply by schakrabortygithub** [reply]
Hi @lindsayad,
This is exactly what I was looking for. Thanks for the suggestion!

Though computation of face material property on this block seems to have some issue, as you have also mentioned. 
I was testing it for a simple elastic problem under uniaxial compression. I output the YY stress on this surface block(PK2_yy_Surface) as well as standard volumetric output(PK2_yy) to Auxvarible of type Constant-Monomial. There is significant difference between the surface values and the volumetric values to the corresponding adjacent element. See the image below,

<img width="1080" alt="moose_SurfaceBlock_MaterialProperty" src="https://github.com/idaholab/moose/assets/35871741/61650e61-d02e-4a30-b613-fa4a9e395392">

My guess is that (not sure though), while computing the material properties on the element-face, the out-of-plane displacement gradients are not getting into the deformation gradient computation. That's causing wrong value of strain and hence stress on element-faces. 

moose input script:

```
[Mesh]
[./gmg]
  type = GeneratedMeshGenerator
  dim = 3
  nx = 4
  ny = 4
  nz = 2
  xmax = 0.1
  ymax = 0.1
  zmax = 0.01
  elem_type = HEX8 #TET4, HEX, HEX8
 []
 [SurfaceBlock]
    input = gmg   # breakmesh surface_FB
    type = LowerDBlockFromSidesetGenerator
    sidesets = 'top bottom left right'
    new_block_id = '2'
    new_block_name = 'surface_LRTB'
    show_info = True
  []
[]

[Outputs]
    file_base = Data_ElasticBasic_n1_Comp
    csv = true
    exodus = true
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Modules/TensorMechanics/Master/all]
  strain = FINITE
  add_variables = true
  #new_system = true
  #formulation = updated
  #volumetric_locking_correction = true
[]

[Materials]
  [./elasticity_tensor_iso]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 1.68e5
    poissons_ratio = 0.25
  [../]
  [./compute_stress]
    type = ComputeFiniteStrainElasticStress
  [../]
[]


[BCs]
  [./BC_All_Y]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = 'top '
    function = '(-0.1)*y*t'
  [../]

  [./BC_BottomLeft_X]
    type = FunctionDirichletBC
    variable = disp_x
    boundary = 'bottom'  # mid_nodes, left_bottom_node back_bottom_node
    function = 0.0
  [../]
   [./BC_BottomLeft_Y]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = 'bottom' # mid_nodes, left_bottom_node
    function = 0.0
  [../]
   [./BC_BottomBack_Z]
    type = FunctionDirichletBC
    variable = disp_z
    boundary = 'bottom'  #'back_bottom_node'
    function = 0.0
  [../]
[]

[Preconditioning]
  [./smp]
    type = SMP
    full = true
  [../]
[]

[Executioner]
    type = Transient
  solve_type = 'PJFNK' #'NEWTON' , 'PJFNK'
  #automatic_scaling = true
  #scaling_group_variables = 'disp_x disp_y disp_z; Rho_EdgePositive Rho_EdgeNegative'
  #petsc_options = '-pc_svd_monitor'
  petsc_options_iname = '-pc_type -pc_asm_overlap -sub_pc_type -ksp_type -ksp_gmres_restart'
  petsc_options_value = ' asm      2              lu            gmres     200'
  nl_abs_tol = 1e-5  #1e-10 for all *_tol
  nl_rel_tol = 1e-5
  nl_abs_step_tol = 1e-5
  nl_max_its = 10
  l_abs_tol = 1e-5
  l_tol = 1e-5
  l_max_its = 20

  dt = 0.05
  dtmin = 0.0001
  dtmax = 0.1
  end_time = 0.2
[]


[AuxVariables]
  [./pk2_yy_Surface]
    order = CONSTANT
    family = MONOMIAL
    block = surface_LRTB
  [../]
  [./pk2_yy]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[AuxKernels]
  [./Aux_pk2_yy_Surface]
   type = RankTwoAux
   variable = pk2_yy_Surface
   rank_two_tensor = stress
   index_j = 1
   index_i = 1
   block = 'surface_LRTB'  
  [../]
  [./Aux_pk2_yy]
   type = RankTwoAux
   variable = pk2_yy
   rank_two_tensor = stress
   index_j = 1
   index_i = 1
  [../]
[]
```

**Reply by GiudGiud** [reply]
@recuero

**Reply by lindsayad** [reply]
@schakrabortygithub I'm adding first class support for this in https://github.com/idaholab/moose/pull/26593. We'll see if it passes this round of testing. Feel free to try it out

**Reply by schakrabortygithub** [reply]
@lindsayad Thanks for the support! I will definitely try it out once that commit(ab0cd0512ed26b6c3a28b93afb358e89d8747e86) becomes available to pull.
